Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1332,6 +1332,66 @@ The following approach can be used in order to patch the JavaScript file:
+
+A well-known use case of this technique is loading the Frida gadget to an application, specially while working on a non-rooted device (this is what [`objection patchapk`](https://github.com/sensepost/objection/wiki/Patching-Android-Applications "Patching Android Applications") basically does).
+
+###### Patching Application's Native Library
+
+Many Android applications use native code in addition to Java code for various performance and security reasons. The native code is present in the form of ELF shared libraries. An ELF executable includes a list of shared libraries (dependencies) that are linked to the executable for it to function optimally. This list can be modified to insert an additional library to be injected into the process.
+
+Modifying the ELF file structure manually to inject a library can be cumbersome and prone to errors. However, this task can be performed with relative ease using [LIEF](https://lief.quarkslab.com/ "LIEF") (Library to Instrument Executable Formats). Using it requires only a few lines of Python code as shown below:
+
+```python
+import lief
+
+libnative = lief.parse("libnative.so")
+libnative.add_library("libinject.so") # Injection!
+libnative.write("libnative.so")
+```
+
+In the above example, libinject.so library is injected as a dependency of a native library (libnative.so), which the application already loads by default. Frida gadget can be injected into an application using this approach as explained in detail in [LIEF's documentation](https://lief.quarkslab.com/doc/latest/tutorials/09_frida_lief.html "LIEF's documentation - How to use frida on a non-rooted device"). As in the previous section, it is important to remember adding the library to the respective architecture `lib` folder in the APK and finally re-signing the application.
+
+###### LD_PRELOAD
+
+Above we looked into techniques which require some kind of modification of the application's code. A library can also be injected into a process using functionalities offered by the loader of the operating system. On Android, which is a Linux based OS, you can load an additional library by setting the `LD_PRELOAD` environment variable.
+
+The symbols loaded from the library passed using LD_PRELOAD always gets the precedence, i.e they are searched first by the loader while resolving the symbols. This feature is often used to modify behaviour of libc functions - fopen(), read(), write(), strcmp() etc, specially in case of obfuscated programs, where information is deliberately obfuscated to prevent static analysis, for instance, file path passed to fopen() or strings passed to strcmp().
```

Comment:
I've restored the original suggestion, as this paragraph does not comply to the style guide.

Apart from that I'm not quite sure about last part.

> [...] information is deliberately obfuscated to prevent static analysis, for instance, file path passed to fopen()

I assume the use case here is: you might be able to patch your program (or not) but definitely you cannot _patch the original libc_. However, you need to get insights on its function parameters. Because `fopen` will open the file but you won't know which file. So you can override `fopen` including a `print`. 

Please take a look and let me know what you think. Gladly we can also discuss it over Slack :)

```suggestion
As the [ld.so man page](http://man7.org/linux/man-pages/man8/ld.so.8.html "LD.SO man page") states, symbols loaded from the library passed using `LD_PRELOAD` always get precedence, i.e they are searched first by the loader while resolving the symbols, effectively overriding the original ones. This feature is often used to inspect the input parameters of some commonly used libc functions such as `fopen`, `read`, `write`, `strcmp`, etc., specially in obfuscated programs, where understanding their behavior may be challenging. Therefore, having an insight on which files are being opened or which strings are being compared may be very valuable. The key idea here is "function wrapping", meaning that you cannot patch system calls e.g. libc's `fopen` but you can override (wrap) it including custom code that will print the input parameters for you but still call the original `fopen`.```